File: api/chalicelib/core/assist.py
import logging
from os import access, R_OK
from os.path import exists as path_exists, getsize

import jwt
import requests
from decouple import config
from fastapi import HTTPException, status

import schemas
from chalicelib.core import projects
from chalicelib.utils.TimeUTC import TimeUTC

logger = logging.getLogger(__name__)

ASSIST_KEY = config("ASSIST_KEY")
ASSIST_URL = config("ASSIST_URL") % ASSIST_KEY


def get_live_sessions_ws_user_id(project_id, user_id):
    data = {
        "filter": {"userId": user_id} if user_id else {}
    }
    return __get_live_sessions_ws(project_id=project_id, data=data)


def get_live_sessions_ws_test_id(project_id, test_id):
    data = {
        "filter": {
            'uxtId': test_id,
            'operator': 'is'
        }
    }
    return __get_live_sessions_ws(project_id=project_id, data=data)


def get_live_sessions_ws(project_id, body: schemas.LiveSessionsSearchPayloadSchema):
    data = {
        "filter": {},
        "pagination": {"limit": body.limit, "page": body.page},
        "sort": {"key": body.sort, "order": body.order}
    }
    for f in body.filters:
        if f.type == schemas.LiveFilterType.METADATA:
            data["filter"][f.source] = {"values": f.value, "operator": f.operator}

        else:
            data["filter"][f.type] = {"values": f.value, "operator": f.operator}
    return __get_live_sessions_ws(project_id=project_id, data=data)


def __get_live_sessions_ws(project_id, data):
    project_key = projects.get_project_key(project_id)
    try:
        results = requests.post(ASSIST_URL + config("assist") + f"/{project_key}",
                                json=data, timeout=config("assistTimeout", cast=int, default=5))
        if results.status_code != 200:
            logger.error(f"!! issue with the peer-server code:{results.status_code} for __get_live_sessions_ws")
            logger.error(results.text)
            return {"total": 0, "sessions": []}
        live_peers = results.json().get("data", [])
    except requests.exceptions.Timeout:
        logger.error("!! Timeout getting Assist response")
        live_peers = {"total": 0, "sessions": []}
    except Exception as e:
        logger.error("!! Issue getting Live-Assist response")
        logger.exception(e)
        logger.error("expected JSON, received:")
        try:
            logger.error(results.text)
        except:
            logger.error("couldn't get response")
        live_peers = {"total": 0, "sessions": []}
    _live_peers = live_peers
    if "sessions" in live_peers:
        _live_peers = live_peers["sessions"]
    for s in _live_peers:
        s["live"] = True
        s["projectId"] = project_id
        if "projectID" in s:
            s.pop("projectID")
    return live_peers


def __get_agent_token(project_id, project_key, session_id):
    iat = TimeUTC.now()
    return jwt.encode(
        payload={
            "projectKey": project_key,
            "projectId": project_id,
            "sessionId": session_id,
            "iat": iat // 1000,
            "exp": iat // 1000 + config("ASSIST_JWT_EXPIRATION", cast=int) + TimeUTC.get_utc_offset() // 1000,
            "iss": config("JWT_ISSUER"),
            "aud": f"openreplay:agent"
        },
        key=config("ASSIST_JWT_SECRET"),
        algorithm=config("JWT_ALGORITHM")
    )


def get_live_session_by_id(project_id, session_id):
    project_key = projects.get_project_key(project_id)
    try:
        results = requests.get(ASSIST_URL + config("assist") + f"/{project_key}/{session_id}",
                               timeout=config("assistTimeout", cast=int, default=5))
        if results.status_code != 200:
            logger.error(f"!! issue with the peer-server code:{results.status_code} for get_live_session_by_id")
            logger.error(results.text)
            return None
        results = results.json().get("data")
        if results is None:
            return None
        results["live"] = True
        results["agentToken"] = __get_agent_token(project_id=project_id, project_key=project_key, session_id=session_id)
    except requests.exceptions.Timeout:
        logger.error("!! Timeout getting Assist response")
        return None
    except Exception as e:
        logger.error("!! Issue getting Assist response")
        logger.exception(e)
        logger.error("expected JSON, received:")
        try:
            logger.error(results.text)
        except:
            logger.error("couldn't get response")
        return None
    return results


def is_live(project_id, session_id, project_key=None):
    if project_key is None:
        project_key = projects.get_project_key(project_id)
    try:
        results = requests.get(ASSIST_URL + config("assistList") + f"/{project_key}/{session_id}",
                               timeout=config("assistTimeout", cast=int, default=5))
        if results.status_code != 200:
            logger.error(f"!! issue with the peer-server code:{results.status_code} for is_live")
            logger.error(results.text)
            return False
        results = results.json().get("data")
    except requests.exceptions.Timeout:
        logger.error("!! Timeout getting Assist response")
        return False
    except Exception as e:
        logger.error("!! Issue getting Assist response")
        logger.exception(e)
        logger.error("expected JSON, received:")
        try:
            logger.error(results.text)
        except:
            logger.error("couldn't get response")
        return False
    return str(session_id) == results


def autocomplete(project_id, q: str, key: str = None):
    project_key = projects.get_project_key(project_id)
    params = {"q": q}
    if key:
        params["key"] = key
    try:
        results = requests.get(
            ASSIST_URL + config("assistList") + f"/{project_key}/autocomplete",
            params=params, timeout=config("assistTimeout", cast=int, default=5))
        if results.status_code != 200:
            logger.error(f"!! issue with the peer-server code:{results.status_code} for autocomplete")
            logger.error(results.text)
            return {"errors": [f"Something went wrong wile calling assist:{results.text}"]}
        results = results.json().get("data", [])
    except requests.exceptions.Timeout:
        logger.error("!! Timeout getting Assist response")
        return {"errors": ["Assist request timeout"]}
    except Exception as e:
        logger.error("!! Issue getting Assist response")
        logger.exception(e)
        logger.error("expected JSON, received:")
        try:
            logger.error(results.text)
        except:
            logger.error("couldn't get response")
        return {"errors": ["Something went wrong wile calling assist"]}
    for r in results:
        r["type"] = __change_keys(r["type"])
    return {"data": results}


def __get_efs_path():
    efs_path = config("FS_DIR")
    if not path_exists(efs_path):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"EFS not found in path: {efs_path}")

    if not access(efs_path, R_OK):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                            detail=f"EFS found under: {efs_path}; but it is not readable, please check permissions")
    return efs_path


def __get_mob_path(project_id, session_id):
    params = {"projectId": project_id, "sessionId": session_id}
    return config("EFS_SESSION_MOB_PATTERN", default="%(sessionId)s") % params


def get_raw_mob_by_id(project_id, session_id):
    efs_path = __get_efs_path()
    path_to_file = efs_path + "/" + __get_mob_path(project_id=project_id, session_id=session_id)
    if path_exists(path_to_file):
        if not access(path_to_file, R_OK):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                                detail=f"Replay file found under: {efs_path};" +
                                       " but it is not readable, please check permissions")
        # getsize return size in bytes, UNPROCESSED_MAX_SIZE is in Kb
        if (getsize(path_to_file) / 1000) >= config("UNPROCESSED_MAX_SIZE", cast=int, default=200 * 1000):
            raise HTTPException(status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, detail="Replay file too large")
        return path_to_file

    return None


def __get_devtools_path(project_id, session_id):
    params = {"projectId": project_id, "sessionId": session_id}
    return config("EFS_DEVTOOLS_MOB_PATTERN", default="%(sessionId)s") % params


def get_raw_devtools_by_id(project_id, session_id):
    efs_path = __get_efs_path()
    path_to_file = efs_path + "/" + __get_devtools_path(project_id=project_id, session_id=session_id)
    if path_exists(path_to_file):
        if not access(path_to_file, R_OK):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                                detail=f"Devtools file found under: {efs_path};"
                                       " but it is not readable, please check permissions")

        return path_to_file

    return None


def session_exists(project_id, session_id):
    project_key = projects.get_project_key(project_id)
    try:
        results = requests.get(ASSIST_URL + config("assist") + f"/{project_key}/{session_id}",
                               timeout=config("assistTimeout", cast=int, default=5))
        if results.status_code != 200:
            logger.error(f"!! issue with the peer-server code:{results.status_code} for session_exists")
            logger.error(results.text)
            return None
        results = results.json().get("data")
        if results is None:
            return False
        return True
    except requests.exceptions.Timeout:
        logger.error("!! Timeout getting Assist response")
        return False
    except Exception as e:
        logger.error("!! Issue getting Assist response")
        logger.exception(e)
        logger.error("expected JSON, received:")
        try:
            logger.error(results.text)
        except:
            logger.error("couldn't get response")
        return False


def __change_keys(key):
    return {
        "PAGETITLE": schemas.LiveFilterType.PAGE_TITLE.value,
        "ACTIVE": "active",
        "LIVE": "live",
        "SESSIONID": schemas.LiveFilterType.SESSION_ID.value,
        "METADATA": schemas.LiveFilterType.METADATA.value,
        "USERID": schemas.LiveFilterType.USER_ID.value,
        "USERUUID": schemas.LiveFilterType.USER_UUID.value,
        "PROJECTKEY": "projectKey",
        "REVID": schemas.LiveFilterType.REV_ID.value,
        "TIMESTAMP": "timestamp",
        "TRACKERVERSION": schemas.LiveFilterType.TRACKER_VERSION.value,
        "ISSNIPPET": "isSnippet",
        "USEROS": schemas.LiveFilterType.USER_OS.value,
        "USERBROWSER": schemas.LiveFilterType.USER_BROWSER.value,
        "USERBROWSERVERSION": schemas.LiveFilterType.USER_BROWSER_VERSION.value,
        "USERDEVICE": schemas.LiveFilterType.USER_DEVICE.value,
        "USERDEVICETYPE": schemas.LiveFilterType.USER_DEVICE_TYPE.value,
        "USERCOUNTRY": schemas.LiveFilterType.USER_COUNTRY.value,
        "PROJECTID": "projectId"
    }.get(key.upper(), key)

File: backend/pkg/assist/proxy/assist.go
package proxy

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"

	"github.com/golang-jwt/jwt/v5"

	"openreplay/backend/internal/config/session"
	"openreplay/backend/pkg/logger"
	"openreplay/backend/pkg/projects"
)

type Assist interface {
	GetLiveSessionByID(projID uint32, sessID uint64) (interface{}, error)
	GetLiveSessionsWS(projID uint32, req *GetLiveSessionsRequest) (interface{}, error)
	IsLive(projID uint32, sessID uint64) (bool, error)
}

type assistImpl struct {
	cfg      *session.Config
	log      logger.Logger
	projects projects.Projects
}

func New(log logger.Logger, cfg *session.Config, projects projects.Projects) (Assist, error) {
	switch {
	case log == nil:
		return nil, errors.New("logger is nil")
	case cfg == nil:
		return nil, errors.New("config is nil")
	case projects == nil:
		return nil, errors.New("projects is nil")
	}
	return &assistImpl{
		log:      log,
		cfg:      cfg,
		projects: projects,
	}, nil
}

// assist.get_live_session_by_id(project_id=project_id, session_id=session_id)
func (a *assistImpl) GetLiveSessionByID(projID uint32, sessID uint64) (interface{}, error) {
	switch {
	case projID == 0:
		return nil, errors.New("projID is 0")
	case sessID == 0:
		return nil, errors.New("sessID is 0")
	}

	proj, err := a.projects.GetProject(projID)
	if err != nil {
		return nil, err
	}
	assistUrl := fmt.Sprintf(a.cfg.AssistUrl, a.cfg.AssistKey) + a.cfg.AssistLiveSuffix + "/" + proj.ProjectKey + "/" + strconv.Itoa(int(sessID))

	resp, err := http.Get(assistUrl)
	if err != nil {
		fmt.Println("Error making request:", err)
		return nil, err
	}
	defer resp.Body.Close()

	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response:", err)
		return nil, err
	}

	// Parse the JSON response
	type assistResponse struct {
		Data map[string]interface{} `json:"data"`
	}
	var response assistResponse
	err = json.Unmarshal(body, &response)
	if err != nil {
		fmt.Println("Error parsing JSON response:", err)
		return nil, err
	}
	if response.Data == nil {
		response.Data = make(map[string]interface{})
	}

	response.Data["live"] = true
	if token, err := a.getAgentToken(projID, proj.ProjectKey, sessID); err != nil {
		a.log.Error(context.Background(), "[proxy] GetLiveSessionByID: ", err)
	} else {
		response.Data["agentToken"] = token
	}
	return response.Data, nil
}

func (a *assistImpl) getAgentToken(projectID uint32, projectKey string, sessionID uint64) (string, error) {
	var method jwt.SigningMethod
	switch a.cfg.AssistJwtAlgorithm {
	case "HS256":
		method = jwt.SigningMethodHS256
	case "HS384":
		method = jwt.SigningMethodHS384
	case "HS512":
		method = jwt.SigningMethodHS512
	default:
		return "", errors.New("unsupported JWT_ALGORITHM: " + a.cfg.AssistJwtAlgorithm)
	}

	// iat in seconds since epoch (UTC), like iat // 1000 in Python
	iat := time.Now().UTC().Unix()
	// local UTC offset in seconds (TimeUTC.get_utc_offset() // 1000 in Python)
	_, offsetSeconds := time.Now().Zone()
	exp := iat + a.cfg.AssistJwtExpiration + int64(offsetSeconds)

	claims := jwt.MapClaims{
		"projectKey": projectKey,
		"projectId":  projectID,
		"sessionId":  fmt.Sprintf("%d", sessionID),
		"iat":        iat,
		"exp":        exp,
		"iss":        a.cfg.AssistJwtIssuer,
		"aud":        "openreplay:agent",
	}

	token := jwt.NewWithClaims(method, claims)
	signed, err := token.SignedString([]byte(a.cfg.AssistJwtSecret))
	if err != nil {
		return "", err
	}
	return signed, nil
}

// assist.get_live_sessions_ws(...)
func (a *assistImpl) GetLiveSessionsWS(projID uint32, req *GetLiveSessionsRequest) (interface{}, error) {
	switch {
	case projID == 0:
		return nil, errors.New("projID is 0")
	case req == nil:
		return nil, errors.New("request body is nil")
	}

	proj, err := a.projects.GetProject(projID)
	if err != nil {
		return nil, err
	}
	assistUrl := fmt.Sprintf(a.cfg.AssistUrl, a.cfg.AssistKey) + a.cfg.AssistLiveSuffix + "/" + proj.ProjectKey

	// Transformation from frontend request to assist request payload
	payload, err := json.Marshal(req.Parse())
	if err != nil {
		return nil, err
	}

	resp, err := a.requestAssistData(assistUrl, payload)
	if err != nil {
		a.log.Error(context.Background(), "Error requesting assist data: %v", err)
		resp = map[string]interface{}{"total": 0, "sessions": []interface{}{}} // default response
	}
	return resp, nil
}

// assist.is_live(project_id=project_id, session_id=session_id, project_key=data["projectKey"])
func (a *assistImpl) IsLive(projID uint32, sessID uint64) (bool, error) {
	switch {
	case projID == 0:
		return false, errors.New("projID is 0")
	case sessID == 0:
		return false, errors.New("sessID is 0")
	}
	proj, err := a.projects.GetProject(projID)
	if err != nil {
		return false, err
	}
	assistUrl := fmt.Sprintf(a.cfg.AssistUrl, a.cfg.AssistKey) + a.cfg.AssistListSuffix + "/" + proj.ProjectKey + "/" + strconv.Itoa(int(sessID))
	resp, err := http.Get(assistUrl)
	if err != nil {
		fmt.Println("error making request:", err)
		return false, err
	}
	defer resp.Body.Close()

	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("error reading response:", err)
		return false, err
	}

	// Parse the JSON response
	var responseData map[string]interface{}
	err = json.Unmarshal(body, &responseData)
	if err != nil {
		fmt.Println("error parsing JSON response:", err)
		return false, err
	}
	return false, nil
}

type GetLiveSessionsRequest struct {
	Filters []interface{} `json:"filters"`
	Sort    string        `json:"sort"`  // "userId", "timestamp" default
	Order   string        `json:"order"` // "asc" or "desc", default "desc"
	Limit   int           `json:"limit"` // default 10
	Page    int           `json:"page"`  // default 1
}

type GetAssistSessionsPayload struct {
	Filter     map[string]interface{} `json:"filter"`
	Pagination map[string]int         `json:"pagination"`
	Sort       map[string]string      `json:"sort"`
}

func (r *GetLiveSessionsRequest) Parse() *GetAssistSessionsPayload {
	if r == nil {
		return nil
	}
	res := &GetAssistSessionsPayload{
		Filter:     make(map[string]interface{}),
		Pagination: map[string]int{"limit": r.Limit, "page": r.Page},
		Sort:       map[string]string{"key": r.Sort, "order": r.Order},
	}
	for _, filter := range r.Filters {
		switch f := filter.(type) {
		case map[string]interface{}:
			filterType := f["type"].(string)
			if filterType == "metadata" {
				filterType = f["source"].(string)
			}
			res.Filter[filterType] = map[string]interface{}{"values": f["value"], "operator": f["operator"]}
		}
	}
	return res
}

/*
The list of fields we have to return:
active
metadata
sessionID
timestamp
userBrowser
userCity
userCountry
userDevice
userDeviceType
userID
userOs
userState
userUUID
*/

func (a *assistImpl) requestAssistData(assistURL string, payload []byte) (interface{}, error) {
	assistReq, err := http.NewRequest("POST", assistURL, bytes.NewBuffer(payload))
	if err != nil {
		fmt.Println("Error creating request:", err)
		return nil, err
	}
	assistReq.Header.Set("Content-Type", "application/json")

	client := &http.Client{
		Timeout: a.cfg.AssistRequestTimeout,
	}
	resp, err := client.Do(assistReq)
	if err != nil {
		fmt.Println("Error sending request:", err)
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		fmt.Printf("HTTP error: %d %s\n", resp.StatusCode, http.StatusText(resp.StatusCode))
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response body:", err)
		return nil, err
	}

	// Parse the JSON response
	var responseData map[string]interface{}
	err = json.Unmarshal(body, &responseData)
	if err != nil {
		fmt.Println("Error parsing JSON response:", err)
		return nil, err
	}

	return responseData, nil
}
